### PR TITLE
Portable scripting

### PR DESCRIPTION
**Pull request by tmat**
Moves all scripting binaries over to FX 4.6 and portable contracts.

Note that this change doesn't make scripting x-plat yet. CoreCLR specific assembly loader needs to be implemented.

### CODE REVIEW

**Comment by tmat** [comment]
@cston @amcasey @KevinH-MS 
FYI @agocke

**Comment by tmat** [comment]
@agocke I'll follow up with cleanup of the contract references - we should have solution-wide list somewhere in settings.targets rather than each project having its own.

File: src/Scripting/CSharp/CSharpObjectFormatter.cs [hunk 1]
```diff
@@ -159,110 +148,18 @@ private static string GetPrimitiveTypeName(SpecialType type)
             }
         }
 
-        private StringBuilder AppendComplexTypeName(StringBuilder builder, Type type, ObjectFormattingOptions options)
+        public override string FormatGeneratedTypeName(Type type)
         {
-            if (type.IsArray)
```
Comment (tmat) [inline src/Scripting/CSharp/CSharpObjectFormatter.cs]: Moved to common.

File: src/Scripting/CSharpTest/ObjectFormatterTests.cs [hunk 2]
```diff
@@ -8,543 +8,18 @@
...
 
-#region Fixtures
-
-
-#pragma warning disable 169 // unused field
-#pragma warning disable 649 // field not set, will always be default value
-namespace ObjectFormatterFixtures
+namespace Microsoft.CodeAnalysis.Scripting.CSharp.UnitTests
 {
-    internal class Outer
```
Comment (tmat) [inline src/Scripting/CSharpTest/ObjectFormatterTests.cs]: Moved to common test project.

File: src/Scripting/Core/InteractiveAssemblyLoader.cs [hunk 3]
```diff
@@ -89,47 +86,71 @@ public override string ToString()
...
+
+            Assembly mscorlib = typeof(object).GetTypeInfo().Assembly;
+            
+            _assembliesLoadedFromLocationByFullPath = new Dictionary<string, AssemblyAndLocation>();
+            _assembliesLoadedFromLocation = new Dictionary<Assembly, LoadedAssembly>();
+            _loadedAssembliesBySimpleName = new Dictionary<string, List<Assembly>>(AssemblyIdentityComparer.SimpleNameComparer);
+            _dependenciesWithLocationBySimpleName = new Dictionary<string, List<AssemblyIdentityAndLocation>>();
 
-            AppDomain.CurrentDomain.AssemblyResolve += new ResolveEventHandler(AssemblyResolve);
+            CorLightup.Desktop.AddAssemblyResolveHandler(AssemblyResolve);
```
Comment (tmat) [inline src/Scripting/Core/InteractiveAssemblyLoader.cs]: An alternative implementation will be needed for CoreCLR.

File: src/Compilers/Core/Portable/CodeAnalysis.csproj [hunk 4]
```diff
@@ -647,6 +646,7 @@
     <InternalsVisibleTo Include="csc" />
     <InternalsVisibleTo Include="csi" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.Scripting" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.Scripting.Desktop" />
```
Comment (amcasey) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: Portable depends on desktop?

File: src/Compilers/Core/Portable/Compilation/EmitResult.cs [hunk 5]
```diff
@@ -13,30 +13,30 @@ public class EmitResult
...
         /// <summary>
         /// True if the compilation successfully produced an executable.
         /// If false then the diagnostics should include at least one error diagnostic
         /// indicating the cause of the failure.
         /// </summary>
-        public bool Success
-        {
-            get { return _success; }
-        }
+        public bool Success => _success;
```
Comment (amcasey) [inline src/Compilers/Core/Portable/Compilation/EmitResult.cs]: Readonly auto-prop?

File: src/Compilers/Core/Portable/CorLightup.cs [hunk 6]
```diff
@@ -48,6 +48,138 @@ private static class CultureTypes
                     return null;
                 }
             }
+
+            private static class _Assembly
+            {
+                internal static readonly Type Type = typeof(Assembly);
+
+                internal static readonly Func<byte[], Assembly> Load_bytes = Type
```
Comment (amcasey) [inline src/Compilers/Core/Portable/CorLightup.cs]: Should any of these be lazy or are we generally going to need all of them?

File: src/Compilers/Core/Portable/CorLightup.cs [hunk 7]
```diff
@@ -48,6 +48,138 @@ private static class CultureTypes
...
+                    .GetTypeInfo()
+                    .GetDeclaredMethod("Load", typeof(byte[]))
+                    .CreateDelegate<Func<byte[], Assembly>>();
```
Comment (amcasey) [inline src/Compilers/Core/Portable/CorLightup.cs]: Consider creating a helper for this series of calls.  It seems to be quite common.

File: src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs [hunk 8]
```diff
@@ -39,7 +39,10 @@ internal sealed class Service : MarshalByRefObject
             private static Control s_ui;
 
             internal static readonly ImmutableArray<string> DefaultSourceSearchPaths =
-                ImmutableArray.Create<string>(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile));
+                ImmutableArray.Create(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile));
```
Comment (amcasey) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs]: Should this be normalized like the ones below?

File: src/Compilers/Core/Portable/CodeAnalysis.csproj (hunk 4 above)
Comment (tmat) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: Leftover. There is no Scripting.Desktop anymore. Will remove.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 9]
```diff
@@ -240,7 +240,7 @@ private void ProcessStarting(InteractiveHostOptions options)
                 preferredCulture: System.Globalization.CultureInfo.CurrentCulture); // TODO (tomat)
 
             _sourceSearchPaths = InteractiveHost.Service.DefaultSourceSearchPaths;
-
+
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: revert?

File: src/Compilers/Core/Portable/Compilation/EmitResult.cs (hunk 5 above)
Comment (tmat) [inline src/Compilers/Core/Portable/Compilation/EmitResult.cs]: yes.

File: src/Compilers/Core/Portable/CorLightup.cs (hunk 7 above)
Comment (tmat) [inline src/Compilers/Core/Portable/CorLightup.cs]: I've followed general pattern we use. I'll keep it as is for now. Andy is working on moving Compiler on CoreCLR portable contracts. once that's done we can clean up all the reflection code.

File: src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs (hunk 8 above)
Comment (tmat) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.Service.cs]: yes.

File: src/Interactive/vbi/My Project/Application.Designer.vb [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ﻿'------------------------------------------------------------------------------
```
Comment (KevinH-MS) [inline src/Interactive/vbi/My Project/Application.Designer.vb]: Can we just delete this file?

File: src/Interactive/vbi/My Project/Settings.Designer.vb [hunk 11]
```diff
@@ -1,7 +1,7 @@
-'------------------------------------------------------------------------------
```
Comment (KevinH-MS) [inline src/Interactive/vbi/My Project/Settings.Designer.vb]: Delete this one too?

File: src/Scripting/CSharp/CSharpObjectFormatter.cs [hunk 12]
```diff
@@ -298,7 +195,7 @@ private string FormatArrayTypeName(Type arrayType, Array arrayOpt, ObjectFormatt
                     for (int i = 0; i < rank; i++)
                     {
                         int lowerBound = arrayOpt.GetLowerBound(i);
-                        long length = arrayOpt.GetLongLength(i);
+                        int length = arrayOpt.GetLength(i);
```
Comment (KevinH-MS) [inline src/Scripting/CSharp/CSharpObjectFormatter.cs]: While I don't often deal with > 2GB arrays, isn't it correct to use LongLength here (or is that not portable)?

File: src/Interactive/vbi/My Project/Settings.Designer.vb (hunk 11 above)
Comment (tmat) [inline src/Interactive/vbi/My Project/Settings.Designer.vb]: I'm not sure if I can delete it and not break VB.

File: src/Scripting/Core/InteractiveAssemblyLoader.cs [hunk 13]
```diff
@@ -17,14 +17,14 @@ namespace Microsoft.CodeAnalysis.Scripting
     /// </summary>
     /// <remarks>
     /// <para>
-    /// An assembly is loaded into CLR's Load Context if it is in the GAC, otherwise it's loaded into No Context via <see cref="Assembly.LoadFile(string)"/>.
-    /// <see cref="Assembly.LoadFile(string)"/> automatically redirects to GAC if the assembly has a strong name and there is an equivalent assembly in GAC.
```
Comment (amcasey) [inline src/Scripting/Core/InteractiveAssemblyLoader.cs]: Consider using verbatim crefs.

File: src/Scripting/CSharp/CSharpObjectFormatter.cs (hunk 12 above)
Comment (tmat) [inline src/Scripting/CSharp/CSharpObjectFormatter.cs]: not portable. I'll ask CoreFx folks what is the support for large arrays.

File: src/Scripting/Core/MetadataShadowCopyProvider.cs [hunk 14]
```diff
@@ -653,21 +635,8 @@ private FileStream CopyFile(string originalPath, string shadowCopyPath, bool fil
             try
             {
                 File.Copy(originalPath, shadowCopyPath, overwrite: true);
```
Comment (KevinH-MS) [inline src/Scripting/Core/MetadataShadowCopyProvider.cs]: When will stuff get cleaned up now?

File: src/Interactive/vbi/My Project/Settings.Designer.vb (hunk 11 above)
Comment (KevinH-MS) [inline src/Interactive/vbi/My Project/Settings.Designer.vb]: I'm *pretty* sure the project system/compiler don't care about it.  It's just generated by the settings designer when you click on the Project Properties page.

File: src/Scripting/Core/MetadataShadowCopyProvider.cs (hunk 14 above)
Comment (tmat) [inline src/Scripting/Core/MetadataShadowCopyProvider.cs]: The host needs to dispose the instance, which it doesn't atm. Fixing.

Comment (tmat) [inline src/Scripting/Core/MetadataShadowCopyProvider.cs]: it currently gets finalized.

File: src/Scripting/Core/ScriptingResources.resx [hunk 15]
```diff
@@ -1,122 +1,103 @@
...
+				</xsd:element>
+			</xsd:choice>
+		</xsd:complexType>
+	</xsd:element>
+</xsd:schema>
+<resheader name="resmimetype">
+	<value>text/microsoft-resx</value>
+</resheader>
+<resheader name="version">
+	<value>1.3</value>
```
Comment (amcasey) [inline src/Scripting/Core/ScriptingResources.resx]: Old value was 2.0.  Was this expected to go backwards?

File: src/Scripting/VisualBasic/VisualBasicObjectFormatter.vb [hunk 16]
```diff
@@ -147,91 +143,24 @@ Namespace Microsoft.CodeAnalysis.Scripting.VisualBasic
...
+        Public Overrides ReadOnly Property GenericParameterClosing As String
+            Get
+                Return ")"
+            End Get
+        End Property
 
-        Public Overrides Function FormatArrayTypeName(array As Array, options As ObjectFormattingOptions) As String
-            Return FormatArrayTypeName(array.[GetType](), array, options)
+        Public Overrides Function FormatGeneratedTypeName(type As Type) As String
+            ' TODO
```
Comment (amcasey) [inline src/Scripting/VisualBasic/VisualBasicObjectFormatter.vb]: TODO?

**Comment by amcasey** [comment]
Superficially LGTM (but it's huge and I might have missed things :)).

File: src/Scripting/VisualBasic/VBScriptingResources.Designer.vb [hunk 17]
```diff
@@ -1,63 +1 @@
...
-            Get
-                Return resourceCulture
-            End Get
-            Set
-                resourceCulture = Value
-            End Set
-        End Property
-    End Module
-End Namespace
+﻿
```
Comment (KevinH-MS) [inline src/Scripting/VisualBasic/VBScriptingResources.Designer.vb]: Delete file if empty?

**Comment by KevinH-MS** [comment]
Naïve question--what aspect of change specifically requires 4.6?

**Comment by KevinH-MS** [comment]
LGTM :+1:

File: src/Scripting/VisualBasic/VisualBasicObjectFormatter.vb (hunk 16 above)
Comment (tmat) [inline src/Scripting/VisualBasic/VisualBasicObjectFormatter.vb]: Will file a workitem

**Comment by tmat** [comment]
The desktop assemblies that reference the portable ones need to resolve the matching facades. Since we are using new portable contracts some of these facades are only present on 4.6

**Pull request merged**